File: scala/5.scala
object Solution {
    def longestPalindrome(s: String): String = {
      def search(pos: Int): (Int, Int) = {
        var l = pos - 1
        var r = pos + 1
        var maxLen = 1
        var res = (pos, pos)
        while (l >= 0 && r < s.length() && s(l) == s(r)){
          if (r-l+1 > maxLen) {
            maxLen = r-l+1
            res = (l, r)
          }
          l -= 1
          r += 1
        }
        l = pos
        r = pos + 1
        while (l>=0 && r < s.length() && s(l) == s(r)) {
          if (r-l+1 > maxLen) {
            maxLen = r-l+1
            res = (l, r)
          }
          l-=1
          r+=1
        }
        res
      }
      var res = ""
      var maxLen = 0
      for ( i <- 0 until s.length()) {
        val range = search(i)
        if (range._2 - range._1 + 1 > maxLen) {
          res = s.substring(range._1, range._2+1)
          maxLen = range._2 - range._1 + 1
        }
      }
      res
    }
}

File: go/5.go
func longestPalindrome(s string) string {
	res := 0
	l, r := 0, 0
	for i := range s {
		ll, rr, res1 := expand(i, i, s, 0)
		if res1 > res {
			res = res1
			l = ll
			r = rr
		}
		ll, rr, res2 := expand(i, i+1, s, 0)
		if res2 > res {
			res = res2
			l = ll
			r = rr
		}
	}
	return s[l : r+1]
}

func expand(l, r int, s string, length int) (int, int, int) {
	if l < 0 || r >= len(s) {
		return l + 1, r - 1, length
	}
	if s[l] == s[r] {
		if l == r {
			return expand(l-1, r+1, s, length+1)
		}
		return expand(l-1, r+1, s, length+2)
	}
	return l + 1, r - 1, length
}